core/capabilities/remote/target/receiver.go

```diff
@@ -0,0 +1,131 @@
+package target
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"sync"
+	"time"
+
+	commoncap "github.com/smartcontractkit/chainlink-common/pkg/capabilities"
+	"github.com/smartcontractkit/chainlink/v2/core/capabilities/remote/types"
+	p2ptypes "github.com/smartcontractkit/chainlink/v2/core/services/p2p/types"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/capabilities"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+)
+
+type remoteTargetReceiver struct {
+	lggr         logger.Logger
+	peerID       p2ptypes.PeerID
+	underlying   commoncap.TargetCapability
+	capInfo      commoncap.CapabilityInfo
+	localDonInfo capabilities.DON
+	workflowDONs map[string]commoncap.DON
+	dispatcher   types.Dispatcher
+
+	requestIDToRequest map[string]*remoteTargetCapabilityRequest
+	requestTimeout     time.Duration
+
+	receiveLock sync.Mutex
+}
+
+var _ types.Receiver = &remoteTargetReceiver{}
+
+func NewRemoteTargetReceiver(ctx context.Context, lggr logger.Logger, peerID p2ptypes.PeerID, underlying commoncap.TargetCapability, capInfo commoncap.CapabilityInfo, localDonInfo capabilities.DON,
+	workflowDONs map[string]commoncap.DON, dispatcher types.Dispatcher, requestTimeout time.Duration) *remoteTargetReceiver {
+
+	receiver := &remoteTargetReceiver{
+		underlying:   underlying,
+		peerID:       peerID,
+		capInfo:      capInfo,
+		localDonInfo: localDonInfo,
+		workflowDONs: workflowDONs,
+		dispatcher:   dispatcher,
+
+		requestIDToRequest: map[string]*remoteTargetCapabilityRequest{},
+		requestTimeout:     requestTimeout,
+
+		lggr: lggr,
+	}
+
+	go func() {
+		timer := time.NewTimer(requestTimeout)
+		defer timer.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-timer.C:
+				receiver.ExpireRequests(ctx)
+			}
+		}
+	}()
+
+	return receiver
+}
+
+func (r *remoteTargetReceiver) ExpireRequests(ctx context.Context) {
+	r.receiveLock.Lock()
+	defer r.receiveLock.Unlock()
+
+	for messageId, executeReq := range r.requestIDToRequest {
+		if time.Since(executeReq.createdTime) > r.requestTimeout {
+
+			if !executeReq.hasResponse() {
+				executeReq.setError(types.Error_TIMEOUT)
+				if err := executeReq.sendResponses(); err != nil {
+					r.lggr.Errorw("failed to send timeout response to all requesters", "capabilityId", r.capInfo.ID, "err", err)
+				}
+			}
+
+			delete(r.requestIDToRequest, messageId)
+		}
+
+	}
+
+}
+
+func (r *remoteTargetReceiver) Receive(msg *types.MessageBody) {
+	// TODO should the dispatcher be passing in a context?
+	ctx := context.Background()
+
+	// TODO Confirm threading semantics of dispatcher receive
+	// TODO May want to have executor per message id to improve liveness
+	r.receiveLock.Lock()
+	defer r.receiveLock.Unlock()
+
+	// TODO multithread this
+
+	if msg.Method != types.MethodExecute {
+		r.lggr.Errorw("received request for unsupported method type", "method", msg.Method)
+		return
+	}
+
+	// A request is uniquely identified by the message id and the hash of the payload
```

Comment:
This sounds risky. A message to a target should be identified by (CallerDonID, WorkflowExecutionID) - available in the metadata. WorkflowExecutionID is something that all nodes in the workflow DON reached consensus on. If we track things only by payload hashes and we have multiple buggy or malicious nodes, it will be very hard for us to make sense of any metrics. And you also need scoping to caller DON.